File: 3660-jump-game-ix/3660-jump-game-ix.py
class Solution:
    def maxValue(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [nums[0]]
        for i in range(1, n):
            res.append(max(res[-1], nums[i]))

        min_idx = n - 1
        for i in range(n - 2, -1, -1):
            if res[i] > nums[min_idx]:
                res[i] = res[min_idx]
            if nums[i] < nums[min_idx]:
                min_idx = i
        return res

File: 3660-jump-game-ix/3660-jump-game-ix.java
class Solution {
    public int[] maxValue(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        res[0] = nums[0];
        for (int i = 1; i < n; i++) {
            res[i] = Math.max(res[i - 1], nums[i]);
        }

        int min_idx = n - 1;
        for (int i = n - 2; i >= 0; i--) {
            if (res[i] > nums[min_idx]) {
                res[i] = res[min_idx];
            }

            if (nums[i] < nums[min_idx]) {
                min_idx = i;
            }
        }
        return res;
    }
}